Firestore/Source/Auth/FSTCredentialsProvider.m

```diff
@@ -74,28 +67,31 @@ - (instancetype)initWithApp:(FIRApp *)app {
   self = [super init];
   if (self) {
     _app = app;
-    _auth = [FIRAuth authWithApp:app];
-    _currentUser = [[FSTUser alloc] initWithUID:self.auth.currentUser.uid];
+    _currentUser = [[FSTUser alloc] initWithUID:[self.app getUID]];
     _userCounter = 0;
 
     // Register for user changes so that we can internally track the current user.
     FSTWeakify(self);
-    _authListenerHandle = [self.auth addAuthStateDidChangeListener:^(FIRAuth *auth, FIRUser *user) {
-      FSTStrongify(self);
-      if (self) {
-        @synchronized(self) {
-          FSTUser *newUser = [[FSTUser alloc] initWithUID:user.uid];
-          if (![newUser isEqual:self.currentUser]) {
-            self.currentUser = newUser;
-            self.userCounter++;
-            FSTVoidUserBlock listenerBlock = self.userChangeListener;
-            if (listenerBlock) {
-              listenerBlock(self.currentUser);
-            }
-          }
-        }
-      }
-    }];
+    _authListenerHandle = [[NSNotificationCenter defaultCenter]
+        addObserverForName:FIRAuthStateDidChangeInternalNotification
+                    object:nil
+                     queue:nil
+                usingBlock:^(NSNotification *notification) {
+                  FSTStrongify(self);
+                  if (self) {
+                    @synchronized(self) {
+                      FSTUser *newUser = [[FSTUser alloc] initWithUID:[self.app getUID]];
```

Comment:
This notification comes back on the main queue and `getUID` blocks waiting on the auth queue so this is probably the wrong place to call this method. One way to deal with this is to redispatch this notification onto the Firestore worker queue, though I think that may be tricky because we block clearing our internal queue waiting on the first user.

The right thing to do here is to change FIRAuth to just provide the user in the notification dictionary. It's already providing the token and it has the user on hand:

https://github.com/firebase/firebase-ios-sdk/blob/343421d93e3b0c9892e1f3d11096d59da8d261fe/Firebase/Auth/Source/FIRAuth.m#L1284

Reply:
Changes to FirebaseAuth:

https://github.com/firebase/firebase-ios-sdk/tree/rsgowman/include_uid_with_internal_auth_state_changes

(will create a PR shortly and then base this PR on that one.)

NB: I only included the uid rather than the entire FIRUser object, figuring that we'd want to avoid importing anything from auth.

Reply:
https://github.com/firebase/firebase-ios-sdk/pull/507

I've rebased this PR on that one, so the commits show up in both. (Will rebase again once 507 is submitted.)